[DCOS-16095] Handle 404 response from Admin Router while waiting for Metronome ser…

adamtheturtle opened the pull request:
…vice.

See https://jira.mesosphere.com/browse/DCOS-16095 - Integration tests: adjust _wait_for_metronome to Admin Router's service endpoint emitting 404 responses

### Testing

I will manually test this now. The PR is marked WIP until then.
I see that there is `dcos_test_utils/test_dcos_api_session.py` but it appears that there isn't really a way to test this easily.
I would appreciate feedback on whether and how to add direct tests for this.

adamtheturtle committed:
Handle 404 response from Admin Router while waiting for Metronome service.

adamtheturtle commented:
@mellenburg and @jgehrcke - This is marked as WIP as I have yet to manually test it. However, if it is urgent and I am asleep, feel free to take a look.

adamtheturtle commented:
I annoyingly have failed to hit this race in my manual testing. I would appreciate a review and guidance on whether and how to test.

adamtheturtle commented:
I reproduced this, and the fix seems to work. It showed the following a few times before moving on:

```
DEBUG:urllib3.connectionpool:https://172.17.0.2:443 "GET /service/metronome/v1/jobs HTTP/1.1" 404 None
INFO:dcos_test_utils.dcos_api_session:It may be the case that Admin Router is returning a 404 despite the Metronome service existing because it uses a cache. This cache is updated periodically.
```

jgehrcke commented:
Thank you, Adam. Let's hold this back for a while until we make progress on DCOS-16108 (sorry, still big thanks for your patch, good solution to the logging challenge!).

vespian commented on dcos_test_utils/dcos_api_session.py:
```diff
@@ -324,11 +324,22 @@ def _wait_for_srouter_slaves_endpoints(self):
-        if r.status_code == 500 or r.status_code == 504:
-            log.info("Metronome gateway timeout, continue waiting for backend...")
+        expected_error_codes = {
+            404: ('It may be the case that Admin Router is returning a 404 '
```
Possible error codes:
503 - cache is borked, this means that Marathon or Mesos are not up yet, continue waiting
404 - metronome has not registered with Mesos yet, continue waiting
502,504 - metronome has registered with mesos, but for some reason, the leading process is not responding, we may want to treat it as an error as I am not sure what can cause such a state. We should probably discuss it a bit.
500 - I do not think this should be a valid state anymore, earlier this was ARs way of saying 404 but it is fixed now.

Hope it helps.

adamtheturtle replied:
@jgehrcke @mellenburg Do you have thoughts on `502,504`?

vespian reviewed (commented):
I wonder if we shouldn't merge it now, as pinpointing DCOS-16108 make take us some time. I run into this error again today: https://teamcity.mesosphere.io/viewLog.html?buildId=678784&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_LaunchAzureResourceMana&tab=buildLog

Thanks for taking care of it @adamtheturtle !

mellenburg commented:
I like the error code messages to help debugging, but why would we want to except Adminrouter errors in the wait_for_metronome step? Wouldn't we want to instead make the `wait_for_adminrouter` function ( https://github.com/mesosphere/dcos-test-utils/blob/master/dcos_test_utils/dcos_api_session.py#L278 ) more robust?

adamtheturtle commented:
Maybe we should wait and replace this whole bunch of code with `docs-checks` if I understand the purpose of https://github.com/dcos/dcos/pull/1609/files.

adamtheturtle committed:
Merge branch 'master' into adminrouter-404-DCOS-16095

jgehrcke reviewed (changes requested):
@mellenburg 

>  but why would we want to except Adminrouter errors in the wait_for_metronome step?

The `wait_for_metronome` is a bad naming choice. The `/service/...` endpoint involves quite a bit of Admin Router logic (since the beginning of 2016) so this has always been tested here, too (implicitly). Now that logic changed a bit. I would like to keep the naming discussion out of the picture for now.

@adamtheturtle as of Pawel's last comment I would love to see the following behavior for all `5xx` responses: have a generic error message for `5xx` HTTP responses, and make sure that the response status code as well as the _body_ is logged in that case. The error/log message emitted by dcos-test-utils should then not make an attempt to explain the failure (as it likely is wrong or unprecise in some regard).

jgehrcke commented:
While DCOS-16108 is being worked on (takes a while, as Pawel pointed out) let's go ahead with this so that dcos-test-utils' wait routine can keep fulfilling its task.

adamtheturtle committed:
Continue waiting on all 5xx status codes

adamtheturtle committed:
Always log the generic message, even if we have a specific message

adamtheturtle committed:
Update expected error codes based on Pawel's feedback

adamtheturtle committed:
Merge branch 'adminrouter-404-DCOS-16095' of github.com:mesosphere/dcos-test-utils into adminrouter-404-DCOS-16095

adamtheturtle committed:
Merge branch 'master' into adminrouter-404-DCOS-16095

adamtheturtle committed:
Comment that we are waiting for admin router

adamtheturtle committed:
Merge branch 'adminrouter-404-DCOS-16095' of github.com:mesosphere/dcos-test-utils into adminrouter-404-DCOS-16095

adamtheturtle commented:
@mhrabovcin @jgehrcke @mellenburg @vespian - I hope that I have addressed comments adequately. As testing this isn't super easy, I'd appreciate a first glance by someone to confirm that this is worth testing.

adamtheturtle commented:
Also see Pawel's request for discussion on https://github.com/mesosphere/dcos-test-utils/pull/30#discussion_r121166012.

jgehrcke commented on dcos_test_utils/dcos_api_session.py:
```diff
@@ -323,12 +323,28 @@ def _wait_for_srouter_slaves_endpoints(self):
+                log.info(error_message)
+            log.info('Continuing to wait for Metronome')
+            log.info('Response body:')
+            log.info(r.content.decode())
```
Is that different from `r.text`?

adamtheturtle replied:
I saw that used elsewhere in this repository, but so is `r.text` so I have changed it.

jgehrcke commented on dcos_test_utils/dcos_api_session.py:
```diff
@@ -323,12 +323,28 @@ def _wait_for_srouter_slaves_endpoints(self):
+            404: ('It may be the case that Admin Router is returning a 404 '
+                  'despite the Metronome service existing because it uses a cache. '
+                  'This cache is updated periodically.'),
+            503: ('The cache is not working. This means that either Marathon '
```
I am not really convinced about the absolute correctness of this error message in all cases where Admin Router emits a 503. In such cases I _strongly_ favor having neutral error messages that have a low risk of providing false/misleading conclusions.

adamtheturtle replied:
@jgehrcke Should I remove just 503 or other errors too?

jgehrcke reviewed (changes requested):
As a quick improvement I'd love us to remove the error message in case of the 503 response, in line with my previous review:

> have a generic error message for 5xx HTTP responses

adamtheturtle committed:
Remove specific message for 503

adamtheturtle committed:
Use response.text rather than response.content.decode()

adamtheturtle commented:
@jgehrcke - ah - I added the generic message for `5xx` errors, but kept the specific ones as well for those where (I thought) we knew the cause.

adamtheturtle commented:
Note that I removed support for 500, as per Pawel's comment.

jgehrcke reviewed (approved):
Nice! :rocket: :ship:

jgehrcke commented:
@mellenburg this should IMO be merged now.

adamtheturtle commented:
ping @mellenburg

mellenburg commented:
@adamtheturtle Sorry for not hearing the ping :grimacing: LGTM

Commit:
Merge branch 'master' into adminrouter-404-DCOS-16095

jgehrcke commented:
What we missed here (CC @vespian) is that we did not emit debugging information in case `if r.status_code in expected_error_codes or r.status_code > 500:` was not `True`.
